### PR TITLE
Android build dependend bluetooth permissions

### PR DESCRIPTION
**Pull request by timkrueger**
- [x] ~Add 'maxSdkVersion' to legacy BLUETOOTH permission~ _This [comment](#issuecomment-1158772212) explains why this is not working._
- [x] Add Android build switch for checking bluetooth permissions
- [x] Request bluetooth permissions on Android 12
  - [ ] ~https://github.com/nextcloud/talk-android/issues/2114~ _This will be solved in #2155._
  - [x] Make initialization of `MagicBluetoothManager` dependend of the permission
- [x] Remove the `Magic` from `MagicAudioManager` and `MagicBluetoothManager`

Resolves: #2132

### CODE REVIEW

**Comment by timkrueger** [comment]
To request the BLUETOOTH_CONNECT permission I must resolve #2114. In that step I try also to get rid of the EffortlessPermissions library and refactor the request permissions workflow.

For an audio call the user will be redirected to the settings app to configure the needed permissions. In an video call the permissions will be requested by dialogs (how it should). I will unify that in this step.

**Comment by mahibi** [comment]
> To request the BLUETOOTH_CONNECT permission I must resolve #2114. In that step I try also to get rid of the EffortlessPermissions library and refactor the request permissions workflow.
> 
> For an audio call the user will be redirected to the settings app to configure the needed permissions. In an video call the permissions will be requested by dialogs (how it should). I will unify that in this step.

from my pov it makes sense to request some permissions additionally after install. So the user don't have to configure any settings when receiving a first call.. 
also see https://github.com/nextcloud/talk-android/issues/1309#issuecomment-900283351

**Comment by timkrueger** [comment]
> from my pov it makes sense to request some permissions additionally after install. So the user don't have to configure any settings when receiving a first call.. also see [#1309 (comment)](https://github.com/nextcloud/talk-android/issues/1309#issuecomment-900283351)

But this can only be additional. The user can  be do the following:

1. Install NC Talk for Android
2. Start the app and give requested permissions
3. Go to settings app and revoke permissions
4. Call comes in and permissions must be requested

or

1. Install NC Talk for Android
2. Start the app and not give requested permissions
3. Call comes in and permissions must be requested

So for usability reasons we can ask for needed permissions during the first start (not during every start). But also when the `CallActivity` is started.

**Comment by timkrueger** [comment]
With all the previous changes now I came into the situation that on Android 12 (SDK 31) the obsolete `BLUETOOTH` permission is needed:

```java
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.nextcloud.talk2, PID: 21710
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.nextcloud.talk2/com.nextcloud.talk.activities.CallActivity}: java.lang.SecurityException: Need BLUETOOTH permission
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
     Caused by: java.lang.SecurityException: Need BLUETOOTH permission
        at android.bluetooth.BluetoothHeadset.<init>(BluetoothHeadset.java:379)
        at android.bluetooth.BluetoothAdapter.getProfileProxy(BluetoothAdapter.java:3017)
        at com.nextcloud.talk.webrtc.MagicBluetoothManager.getBluetoothProfileProxy(MagicBluetoothManager.java:314)
        at com.nextcloud.talk.webrtc.MagicBluetoothManager.start(MagicBluetoothManager.java:155)
        at com.nextcloud.talk.webrtc.MagicAudioManager.startBluetoothManager(MagicAudioManager.java:120)
        at com.nextcloud.talk.webrtc.MagicAudioManager.start(MagicAudioManager.java:236)
        at com.nextcloud.talk.activities.CallActivity.basicInitialization(CallActivity.java:463)
        at com.nextcloud.talk.activities.CallActivity.onCreate(CallActivity.java:341)
        at android.app.Activity.performCreate(Activity.java:8051)
        at android.app.Activity.performCreate(Activity.java:8031)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.app.ActivityThread.main(ActivityThread.java:7839) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```

After taking a look with @tobiasKaminsky (Thank you 💙 ) we understood what the problem is. 


During runtime (on Android 12) while starting the `MagicBluetoothManger` the following part in `android.bluetooth.BluetootHeadset` constructor will be executed and results in the previous exception:

```java
        // Preserve legacy compatibility where apps were depending on
        // registerStateChangeCallback() performing a permissions check which
        // has been relaxed in modern platform versions
        if (context.getApplicationInfo().targetSdkVersion <= Build.VERSION_CODES.R
                && context.checkSelfPermission(android.Manifest.permission.BLUETOOTH)
                        != PackageManager.PERMISSION_GRANTED) {
            throw new SecurityException("Need BLUETOOTH permission");
        }
```

In the `build.gradle` the `targetSdkVersion 30` and `compileSdkVersion 31` is configured. In the `AndroidManifest.xml` the `BLUETOOTH` permission with the attribute `android:maxSdkVersion="30"` is configured.

Because the `MagicBluetoothManager` checks for the `targetSdkVersion` instead of `Build.VERSION.SDK_INT` also on a Android 12 (SDK 31) the `BLUETOOTH` permission is needed.

So the solution will be to remove the `android:maxSdkVersion` for the `BLUETOOTH` permission and additionally request the `BLUETOOTH_CONNECT` permission.

**Comment by AndyScherzinger** [comment]
> Please can you test this with Bluetooth devices on your Android 12 device and report here if you got a permission request!? Surely your Bluetooth device should also work

Just tested on my OP9Pro: no permission request, no bt device listed for audio, no bt permission in the app management visible

**Comment by maxcco** [comment]
> For some to me unknown reasons for me it is not longer possible to get the request for the `BLUETOOTH_CONNECT` permission in the Android emulator or my test device on Android 12.
> 
> Please can you test this with Bluetooth devices on your Android 12 device and report here if you got a permission request!? Surely your Bluetooth device should also work 😉
> 
> /cc @AndyScherzinger , @AlvaroBrey , @tobiasKaminsky , @mahibi

I can help with testing (since I reported the problem), the link you provide doesn't work for me though

**Review by mahibi — APPROVED** [review]

**Comment by mahibi** [comment]
tested bluetooth headset, picture in picture and notifications with android 10 and 12.
everything worked as expected :+1:

**Comment by AndyScherzinger** [comment]
Tested on Android 12 with connected bt speaker, works fine requesting permissions initially and after having revoked them 👍

**Review by AndyScherzinger — APPROVED** [review]

**Comment by timkrueger** [comment]
> I can help with testing (since I reported the problem), the link you provide doesn't work for me though

We'll create a RC today. It would be great if you will test that. :-)

File: app/src/main/java/com/nextcloud/talk/webrtc/WebRtcBluetoothManager.java [hunk 1]
```diff
@@ -47,20 +50,23 @@
...
 
 import java.util.List;
 import java.util.Set;
 
-public class MagicBluetoothManager {
-    private static final String TAG = "MagicBluetoothManager";
+import androidx.core.app.ActivityCompat;
+
+public class WebRtcBluetoothManager {
+    private static final String TAG = WebRtcBluetoothManager.class.getCanonicalName();
```
Comment (AlvaroBrey) [inline app/src/main/java/com/nextcloud/talk/webrtc/WebRtcBluetoothManager.java:62]: should be `getSimpleName()` instead, as `canonicalName` can easily go over the tag character limit.

File: app/src/gplay/java/com/nextcloud/talk/services/firebase/ChatAndCallMessagingService.kt [hunk 2]
```diff
@@ -214,19 +214,23 @@ class MagicFirebaseMessagingService : FirebaseMessagingService() {
...
-                    this@MagicFirebaseMessagingService,
+                    this@ChatAndCallMessagingService,
                     0,
                     fullScreenIntent,
-                    PendingIntent.FLAG_UPDATE_CURRENT
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                        PendingIntent.FLAG_MUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
+                    } else {
+                        PendingIntent.FLAG_UPDATE_CURRENT
+                    }
```
[the remark this answers is not in the document]
Comment (timkrueger) [reply app/src/gplay/java/com/nextcloud/talk/services/firebase/ChatAndCallMessagingService.kt:224]: You're right. I can be immutable.

**Pull request merged**